cache-then-network/index.js

```diff
@@ -0,0 +1,188 @@
+  var shouldCacheError = cacheFailInput.checked;
+  if (shouldCacheError || !res) {
+    throw Error('Cache miss');
+  }
+
+  // Read the response as json
+  res.json().then(function(data) {
+    // Only update the page if the network request hasn't
+    // already returned. Otherwise we'd be overwriting the
+    // data we just pulled from the network.
+    if (!gotNetworkData) {
+      updatePage(data);
+    }
+  });
+}
+
+
+/*
+ * This is the heart of the demo. Whenever the user clicks the
+ * button, we'll make simultaneous (or nearly-simultaneous) requests
+ * of the cache and the network for data. In a real application, this
+ * would probably happen on page load rather than as a result of
+ * user action, but to make the page more illustrative and allow
+ * for user-specified delays, we've got a button.
+ */
+getDataButton.addEventListener('click', function handleClick() {
+  disableUI();
+
+  // Initiate network fetch
+  networkStatus.textContent = 'Fetching...';
+  networkFetchStartTime = window.performance.now();
+  // Use the Fetch API to actually request data
+  var networkFetch = fetch(dataUrl, { mode: 'cors', cache: 'no-cache' }).then(function(res) {
+    var networkDelay = networkDelayInput.value || 0;
+
+    // We simulate network delays by waiting before actually calling
+    // our network fetch callback. If the callback errors, we want
+    // to make sure to reject the Promise we got from the original
+    // fetch.
+    return new Promise(function(resolve, reject) {
+      setTimeout(function() {
+        try {
+          handleFetchCompletion(res);
+          resolve();
+        } catch (err) {
+          reject(err);
+        }
+      }, networkDelay);
+    });
+  }).then(function() {
+    var now = window.performance.now();
+    var elapsed = Math.round(now - networkFetchStartTime);
+    networkStatus.textContent = 'Success after ' + elapsed + 'ms';
+  }).catch(function(err) {
+    var now = window.performance.now();
+    var elapsed = Math.round(now - networkFetchStartTime);
+    networkStatus.textContent = err + ' after ' + elapsed + 'ms';
+  });
+
+  // Get cached data
+  cacheStatus.textContent = 'Fetching...';
+  cacheFetchStartTime = window.performance.now();
+  var cacheFetch = caches.open(cacheName).then(function(cache) {
+    return cache.match(dataUrl).then(function(res) {
+      var cacheDelay = cacheDelayInput.value || 0;
+
+      // We simulate cache delays by waiting before actually calling
+      // our cache fetch callback. If the callback errors, we want
+      // to make sure to reject the Promise we got from the original
+      // call to `match`.
+      return new Promise(function(resolve, reject) {
+        setTimeout(function() {
+          try {
+            handleCacheFetchCompletion(res);
+            resolve();
+          } catch (err) {
+            reject(err);
+          }
+        }, cacheDelay);
+      });
+    }).then(function() {
+      var now = window.performance.now();
```

Comment:
If you use `Date.now()` you would not need to round here. Do you see sub-millisecond results?